core/domain/exp_domain.py

```diff
@@ -2964,6 +2967,82 @@ def delete_state(self, state_name):
 
         del self.states[state_name]
 
+
+    def get_state_names_mapping(self, change_list):
+        """Creates a mapping from the current state names of the exploration
+        to their corresponding state names of the older version.
+
+        Args:
+            change_list: list(dict). A list of changes introduced in this
+                commit.
+
+        Returns:
+            dict. The new_to_old_state_names mapping dict.
+        """
+        old_to_new_state_names = {}
+
+        for state_name in self.states:
+            old_to_new_state_names[state_name] = state_name
+
+        for change_dict in reversed(change_list):
+            if change_dict['cmd'] == CMD_RENAME_STATE:
+                old_to_new_state_names[change_dict['old_state_name']] = (
+                    old_to_new_state_names.pop(change_dict['new_state_name']))
+
+        new_to_old_state_names = {
+            value: key for key, value in old_to_new_state_names.items()
```

Comment:
FYI, you can also do `dict(zip(old_to_new_state_names.values(), old_to_new_state_names.keys()))`. No need to follow this. Just wanted to let you know.